### DISCUSSION TITLE
juicefs client out of memory

### DISCUSSION THREAD

**Discussion by deelaaay** [juicedata/juicefs#1529] [General]
Out of memory: Kill process 2523920 (juicefs) score 444 or sacrifice child
Killed process 2523920 (juicefs) total-vm:15826948kB, anon-rss:14931252kB, file-rss:0kB, shmem-rss:0kB
oom_reaper: reaped process 2523920 (juicefs), now anon-rss:0kB, file-rss:0kB, shmem-rss:0kB

exec  dmesg  ,display the above content。 how do I control the JuiceFS client  physical memory usage?

**Comment by davies** [comment]
There is a option (`--buffer-size`) to control the memory used for read/write, it's 300MB by default, the actual used memory could be higher than that, but should not be as high as 15GB.

What's the workload were you running? Which version?

**Reply by deelaaay** [reply]
juicefs stats /mount-point --verbosity 1
![image](https://user-images.githubusercontent.com/22955222/157788243-e694ac3a-9d4c-480d-8b1f-ef823f27776b.png)


 juicefs --version
juicefs version 1.0.0-beta1 (2022-01-13T15:45:45Z 28a42cf)

workload were videos and pictures，as shown above picture，The buffer size is always less than 300M。 but the  juicefs client process used memory  keeps growing until  oom 。 how do I control the JuiceFS client physical memory usage?

**Reply by davies** [reply]
There could be a bug causing memory leaking, can you upload the heap of the process:

```
wget localhost:6060/debug/pprof/heap
```

If the above command fail, please find the correct port of debug agent using `lsof`

**Reply by deelaaay** [reply]
[root@node165 ~]# lsof -i -nP | grep LISTEN | grep juicefs
juicefs   133460     root   11u  IPv4  410700      0t0  TCP 127.0.0.1:6062 (LISTEN)
juicefs   133460     root   13u  IPv4  408878      0t0  TCP 127.0.0.1:6072 (LISTEN)
juicefs   133460     root   17u  IPv4  405295      0t0  TCP 127.0.0.1:9567 (LISTEN)

[root@node165 ~]# wget localhost:6062/debug/pprof/heap

[heap.zip](https://github.com/juicedata/juicefs/files/8228604/heap.zip)

thanks

**Reply by davies** [reply]
Here are the graph generated by `go tool pprof`

![profile002](https://user-images.githubusercontent.com/40902/157798395-e1c8366b-bbda-4b90-b195-b1420d1c6abe.svg)
![profile001](https://user-images.githubusercontent.com/40902/157798400-57e69e1e-4d71-457f-ae06-06ac8e109e6e.svg)

We can see there are lots of goroutines (2.7 millions) waiting to upload the blocks, and lots of objects used by goroutines are not freed. So it's caused by slow upload of objects in writeback mode, we will optimized this use case and terminate these goroutines to save memory.

**Reply by SandyXSD** [reply]
How many files in your volume (df -i)? And what's the metadata engine? It may be caused by the backup of metadata.
Mount with `-v` enables [debug log](https://github.com/juicedata/juicefs/blob/v1.0.0-beta1/pkg/vfs/backup.go#L63) for the start time of backup.
